lib/AST/Type.cpp

```diff
@@ -4089,6 +4089,58 @@ Type TypeBase::openAnyExistentialType(ArchetypeType *&opened) {
 }
 
 // SWIFT_ENABLE_TENSORFLOW
+Type TypeBase::getAutoDiffAssociatedType(
+    AutoDiffAssociatedTypeKind kind, LookupConformanceFn lookupConformance) {
+  assert(lookupConformance);
+  auto &ctx = getASTContext();
+
+  // Builtins floats are their own Tangent/Cotangent.
+  if (is<BuiltinType>())
+    return this;
+
+  // Tuples' Tangent/Cotangent is a tuple of each element's Tangent/Cotangent.
+  if (auto *tupleTy = getAs<TupleType>()) {
+    SmallVector<TupleTypeElt, 8> newElts;
+    for (auto elt : tupleTy->getElements()) {
+      auto eltAssocTy =
+          elt.getType()->getAutoDiffAssociatedType(kind, lookupConformance);
+      if (!eltAssocTy)
+        return Type();
+      newElts.push_back(elt.getWithType(eltAssocTy));
+    }
+    return TupleType::get(newElts, ctx);
+  }
+
+  // Find the TangentVector/CotangentVector associated type on the
+  // Differentiable protocol.
+  auto *differentiableProtocol =
+      ctx.getProtocol(KnownProtocolKind::Differentiable);
+  assert(differentiableProtocol && "could not find differentiable protocol");
+  Identifier associatedTypeIdentifier;
+  switch (kind) {
+  case AutoDiffAssociatedTypeKind::TangentVector:
+    associatedTypeIdentifier = ctx.getIdentifier("TangentVector");
```

Comment:
I'd add these to `KnownIdentifiers.def`.

Reply:
will do